### DISCUSSION TITLE
ESP32 and ulab - firmware

### DISCUSSION THREAD

**Discussion by marestudio2004** [micropython/micropython#10473] [ESP32]
Hello everyone,

on my ESP32 I'm trying to run a project with:

https://github.com/mocleiri/tensorflow-micropython-examples/tree/main/examples/micro-speech

on the microcontroller I installed the latest version of firmware from https://micropython.org/download/esp32/. 

The problem is that this version does not have the ulab library!

I tried in every way to generate the firmware with ulab myself, but I couldn't succeed!

Does anyone have experience and can recommend the fastest way to solve this problem (downloading the firmware in ...)?

**Comment by 2dof** [comment]
Yu need mkropython with in-build ulab idf libraries,  see:
https://github.com/v923z/micropython-ulab/ 


On:    https://gitlab.com/rcolistete/micropython-samples/-/tree/master/ESP32/Firmware/v1.12_with_ulab/ulab_v0.54.0_2020-07-29

You will find already compiled binaries to downland,  then by using esptool just flash your esp32.
flashing with esptool :  https://docs.espressif.com/projects/esptool/en/latest/esp32/esptool/flashing-firmware.html
                                     https://docs.micropython.org/en/latest/esp32/tutorial/intro.html#deploying-the-firmware


for example  ( in my case: esp32 wroom 32, connected to COMport 8, ) : 
commands : 
```
 >> python -m esptool --port COM8 erase_flash
 >> python -m esptool --chip esp32 --port COM8 write_flash -z 0x1000 esp32_idf4_ulab_sp_thread_v1.12-663-gea4670d5a_2020-07-29.bin
```

**Reply by marestudio2004** [reply]
Thanks!

**Reply by v923z** [reply]
@marestudio2004 If you fixed the issue, would you mind submitting a PR here: https://github.com/v923z/micropython-builder? That would generate the firmware automatically.
